### PR TITLE
feat: search for flox.nix at root

### PR DESCRIPTION
**Pull request by tomberek**
CLI commands do not understand these yet

### CODE REVIEW

File: plugins/rootFloxEnvs.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  lib,
+  self,
+  ...
+}: {
+  sourceType,
```
Comment (mkenigs) [inline plugins/rootFloxEnvs.nix:6]:
```suggestion
```

File: plugins/rootFloxEnvs.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  injectedArgs ? {},
+  dir ? sourceType,
```
Comment (mkenigs) [inline plugins/rootFloxEnvs.nix:8]:
```suggestion
```

File: plugins/rootFloxEnvs.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+}: {context, ...}: let
+  floxEnvDir = context.self.outPath;
+  floxNixPath = "${floxEnvDir}/flox.nix";
+  catalogPath = "${floxEnvDir}/catalog.json";
+  namespace = ["default"];
+  result = lib.genAttrs context.systems (system: {
```
Comment (mkenigs) [inline plugins/rootFloxEnvs.nix:14]: This feels like not the capacitor way of doing this. @ysndr might have more input?

**Review by mkenigs — CHANGES_REQUESTED** [review]
If we're going down this route for sure, I think rather than having a separate plugin, we should consider modifying our existing plugin to accept the root flake dir as an argument. The added benefit is that we could drop the unnecessary `default.nix`

File: plugins/rootFloxEnvs.nix (hunk 3 above)
Comment (ysndr) [reply plugins/rootFloxEnvs.nix:14]: floxEnvs/flox.nix dont comply with the capacitor interface (...) thus we have to import them explicitly

**Comment by ysndr** [comment]
> If we're going down this route for sure, I think rather than having a separate plugin, we should consider modifying our existing plugin to accept the root flake dir as an argument. The added benefit is that we could drop the unnecessary default.nix

Since we want to support both at root and in a subtree, i think its easier to just do it as a separate plugin.

**Comment by mkenigs** [comment]
> Since we want to support both at root and in a subtree, i think its easier to just do it as a separate plugin. We still cant drop the `default.nix`s because we rely on them to find the flox.nix files in the first place

That's kinda what I meant we should change - we should stop relying on `packages` to find `flox.nix`. And if we have a plugin that can load `flox.nix` files directly, it seems like loading a `flox.nix` from the root should be trivial

**Comment by ysndr** [comment]
> we should stop relying on packages to find flox.nix

flox develop does still rely on a an associated package for a flox.nix, (or an associated flox.nix for the package). loading `flox.nix`s independently makes their correspondence by namespace only incidentally

**Comment by mkenigs** [comment]
`flox develop` could check if a package of the same name exists, and only call `nix develop` if it does

I think having free floating `flox.nix` files is the direction @garbas is thinking is better for UX

**Comment by ysndr** [comment]
> flox develop could check if a package of the same name exists, and only call nix develop if it does

yes that's the point about incidentally I mean
sure enough its a weird UX that you have to have some mock default.nix just to have the flox env available.
I already find it weird that they are connected to packages in the first place actually..
I'd rather have flox develop checkout `<package>.meta.floxEnv` (defaults to `context.self.floxEnvs.<system>.<namespace...> or false`) i.e. make that connection explicit rather than having flox decide heuristically.

**Comment by mkenigs** [comment]
I could see that being more clear, although all it's doing is moving logic from Rust to Nix right? Either way we'll have to check whether something exists in `.#floxEnvs.$system.namespace`

**Comment by ysndr** [comment]
> although all it's doing is moving logic from Rust to Nix right?

Yes but in nix, (power) users have control over it (including do unholy things such as binding a floxenv from a different repo 🙊 and sharing envs between packages)

**Comment by mkenigs** [comment]
🙈

**Pull request merged**